File: android/app/src/main/kotlin/com/poracode/app/chat/RichMediaPolicy.kt
package com.poracode.app.chat

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

enum class RichInlineImageKind {
    DATA_URL,
    RAW_SVG,
    BASE64,
}

data class RichInlineImageClassification(
    val kind: RichInlineImageKind,
    val mimeType: String,
)

sealed interface RichImagePathPart {
    data class Key(val value: String) : RichImagePathPart
    data class Index(val value: Long) : RichImagePathPart
}

data class RichRemoteImageRef(
    val threadId: String,
    val itemId: String,
    val path: List<RichImagePathPart>,
    val mimeType: String,
    val bytes: Long,
    val width: Long? = null,
    val height: Long? = null,
    val preview: String? = null,
)

data class RichOmittedPayload(val bytes: Long)

object RichImagePolicy {
    private val signatures = listOf(
        "iVBORw0KGgo" to "image/png",
        "/9j/" to "image/jpeg",
        "R0lGOD" to "image/gif",
        "UklGR" to "image/webp",
        "PHN2Zw" to "image/svg+xml",
        "PD94bWwg" to "image/svg+xml",
    )
    private val resultStringKeys = listOf(
        "dataUrl",
        "data_url",
        "image",
        "b64_json",
        "base64",
        "png",
        "data",
        "src",
        "content",
        "text",
    )
    private val resultArrayKeys = listOf("images", "data", "content", "output")

    /** Only self-contained image bytes are displayable; URLs and local schemes are rejected. */
    fun classify(source: String): RichInlineImageClassification? {
        val head = source.take(16).trimStart()
        if (Regex("^data:image/", RegexOption.IGNORE_CASE).containsMatchIn(head)) {
            return RichInlineImageClassification(RichInlineImageKind.DATA_URL, dataUrlMime(source))
        }
        if (Regex("^<svg(?:\\s|>)", RegexOption.IGNORE_CASE).containsMatchIn(head) ||
            Regex("^<\\?xml", RegexOption.IGNORE_CASE).containsMatchIn(head)
        ) {
            return RichInlineImageClassification(RichInlineImageKind.RAW_SVG, "image/svg+xml")
        }
        for ((prefix, mime) in signatures) {
            if (source.startsWith(prefix)) {
                return RichInlineImageClassification(RichInlineImageKind.BASE64, mime)
            }
        }
        return null
    }

    fun decodeRemoteRef(value: JsonElement?): RichRemoteImageRef? {
        val marker = value?.objectOrNull()?.get("__poracodeImageRef")?.objectOrNull() ?: return null
        val threadId = marker.requiredString("threadId", allowEmpty = false) ?: return null
        val itemId = marker.requiredString("itemId", allowEmpty = false) ?: return null
        val mime = marker.requiredString("mime")?.takeIf { it.startsWith("image/") } ?: return null
        val bytes = marker["bytes"]?.longOrStrictNull()?.takeIf { it >= 0 } ?: return null
        val pathRaw = marker["path"]?.arrayOrNull()?.takeIf { it.isNotEmpty() } ?: return null
        val path = pathRaw.map { part ->
            part.stringOrNull()?.let(RichImagePathPart::Key)
                ?: part.longOrStrictNull()?.let(RichImagePathPart::Index)
                ?: return null
        }
        val width = optionalNonNegativeLong(marker, "width") ?: return null
        val height = optionalNonNegativeLong(marker, "height") ?: return null
        val preview = marker.optionalString("preview")
        if (preview is RichField.Invalid) return null
        val previewValue = preview.valueOrNull()
        if (previewValue != null &&
            !Regex("^data:image/", RegexOption.IGNORE_CASE).containsMatchIn(previewValue)
        ) {
            return null
        }
        return RichRemoteImageRef(
            threadId,
            itemId,
            path,
            mime,
            bytes,
            width.value,
            height.value,
            previewValue,
        )
    }

    fun decodeOmitted(value: JsonElement?): RichOmittedPayload? {
        val marker = value?.objectOrNull()?.get("__poracodeOmitted")?.objectOrNull() ?: return null
        val bytes = marker["bytes"]?.longOrStrictNull()?.takeIf { it >= 0 } ?: return null
        return RichOmittedPayload(bytes)
    }

    fun contentBlockIsSafe(block: RichContentBlock.Image): Boolean =
        classify(block.dataUrl) != null

    fun hasDisplayableImage(payload: JsonObject?): Boolean {
        if (payload == null || payload["status"]?.stringOrNull() == "error") return false
        val images = payload["images"] as? JsonArray
        if (images?.any(::isSafeCandidate) == true) return true
        val result = payload["result"] ?: return false
        if (isSafeCandidate(result)) return true
        val resultObject = result as? JsonObject ?: return false
        for (key in resultStringKeys) {
            if (resultObject[key]?.let(::isSafeCandidate) == true) return true
        }
        for (key in resultArrayKeys) {
            val values = resultObject[key] as? JsonArray ?: continue
            for (entry in values) {
                if (isSafeCandidate(entry)) return true
                val nested = entry as? JsonObject ?: continue
                if (resultStringKeys.any { nested[it]?.let(::isSafeCandidate) == true }) return true
            }
        }
        return false
    }

    private data class OptionalLongValue(val value: Long?)

    private fun optionalNonNegativeLong(value: JsonObject, key: String): OptionalLongValue? {
        return when (val field = value.optionalLong(key)) {
            RichField.Missing -> OptionalLongValue(null)
            RichField.Invalid -> null
            is RichField.Value -> field.value.takeIf { it >= 0 }?.let(::OptionalLongValue)
        }
    }

    private fun isSafeCandidate(value: JsonElement): Boolean =
        value.stringOrNull()?.let(::classify) != null || decodeRemoteRef(value) != null

    private fun dataUrlMime(source: String): String {
        val match = Regex("^\\s*data:([^;,]+)[;,]", RegexOption.IGNORE_CASE).find(source)
        val mime = match?.groupValues?.getOrNull(1)?.lowercase()
        return mime?.takeIf { it.startsWith("image/") } ?: "image/png"
    }
}

enum class RichAttachmentError(val wireCode: String) {
    EMPTY("empty_attachment"),
    TOO_LARGE("attachment_too_large"),
    INVALID("invalid_attachment"),
}

data class RichAttachmentDecision(
    val queryValid: Boolean,
    val bodyWithinLimit: Boolean,
    val accepted: Boolean,
    val error: RichAttachmentError?,
)

object RichAttachmentPolicy {
    const val MAX_BODY_BYTES: Long = 20L * 1024L * 1024L
    const val MAX_NAME_UTF16_UNITS: Int = 255

    fun evaluate(name: String, byteCount: Long): RichAttachmentDecision {
        val queryValid = name.isNotEmpty() && name.length <= MAX_NAME_UTF16_UNITS
        val bodyWithinLimit = byteCount in 0..MAX_BODY_BYTES
        val accepted = queryValid && bodyWithinLimit && byteCount > 0
        val error = when {
            accepted -> null
            !queryValid || byteCount < 0 -> RichAttachmentError.INVALID
            !bodyWithinLimit -> RichAttachmentError.TOO_LARGE
            else -> RichAttachmentError.EMPTY
        }
        return RichAttachmentDecision(queryValid, bodyWithinLimit, accepted, error)
    }
}

File: ios/App/App/Features/RichChat/RichMediaPolicy.swift
import Foundation

enum RichInlineImageKind: Sendable, Equatable {
  case dataURL
  case rawSVG
  case base64
}

struct RichInlineImageClassification: Sendable, Equatable {
  let kind: RichInlineImageKind
  let mimeType: String
}

enum RichImagePathPart: Sendable, Equatable {
  case key(String)
  case index(Int64)
}

struct RichRemoteImageReference: Sendable, Equatable {
  let threadID: String
  let itemID: String
  let path: [RichImagePathPart]
  let mimeType: String
  let bytes: Int64
  let width: Int64?
  let height: Int64?
  let preview: String?
}

struct RichOmittedPayload: Sendable, Equatable {
  let bytes: Int64
}

enum RichImagePolicy {
  private static let signatures: [(String, String)] = [
    ("iVBORw0KGgo", "image/png"),
    ("/9j/", "image/jpeg"),
    ("R0lGOD", "image/gif"),
    ("UklGR", "image/webp"),
    ("PHN2Zw", "image/svg+xml"),
    ("PD94bWwg", "image/svg+xml"),
  ]

  /// Allows only self-contained image bytes. Remote/local/script URLs are never displayable.
  static func classify(_ source: String) -> RichInlineImageClassification? {
    let trimmed = source.drop(while: { $0.isWhitespace })
    let head = String(trimmed.prefix(16)).lowercased()
    if head.hasPrefix("data:image/") {
      let metadata = trimmed.prefix { $0 != "," }
      let mime = metadata.dropFirst(5).prefix { $0 != ";" }.lowercased()
      return RichInlineImageClassification(
        kind: .dataURL,
        mimeType: mime.hasPrefix("image/") ? String(mime) : "image/png"
      )
    }
    if head.hasPrefix("<svg") || head.hasPrefix("<?xml") {
      return RichInlineImageClassification(kind: .rawSVG, mimeType: "image/svg+xml")
    }
    for (prefix, mime) in signatures where source.hasPrefix(prefix) {
      return RichInlineImageClassification(kind: .base64, mimeType: mime)
    }
    return nil
  }

  static func decodeRemoteReference(_ value: RichJSON?) -> RichRemoteImageReference? {
    guard let marker = value?.objectValue?["__poracodeImageRef"]?.objectValue,
      let threadID = RichDecoding.requiredString(marker, "threadId", allowEmpty: false),
      let itemID = RichDecoding.requiredString(marker, "itemId", allowEmpty: false),
      let mime = RichDecoding.requiredString(marker, "mime"), mime.hasPrefix("image/"),
      let bytes = marker["bytes"]?.exactInt64Value, bytes >= 0,
      let rawPath = marker["path"]?.arrayValue, !rawPath.isEmpty
    else { return nil }

    var path: [RichImagePathPart] = []
    for part in rawPath {
      if let key = part.stringValue {
        path.append(.key(key))
      } else if let index = part.exactInt64Value, index >= 0 {
        path.append(.index(index))
      } else {
        return nil
      }
    }
    guard let width = optionalNonnegativeInteger(marker, "width"),
      let height = optionalNonnegativeInteger(marker, "height")
    else { return nil }
    let preview = RichDecoding.optionalString(marker, "preview")
    guard preview != .invalid else { return nil }
    if let source = preview.value {
      guard
        source.trimmingCharacters(in: .whitespacesAndNewlines)
          .lowercased().hasPrefix("data:image/")
      else { return nil }
    }
    return RichRemoteImageReference(
      threadID: threadID,
      itemID: itemID,
      path: path,
      mimeType: mime,
      bytes: bytes,
      width: width.value,
      height: height.value,
      preview: preview.value
    )
  }

  static func decodeOmitted(_ value: RichJSON?) -> RichOmittedPayload? {
    guard let marker = value?.objectValue?["__poracodeOmitted"]?.objectValue,
      let bytes = marker["bytes"]?.exactInt64Value, bytes >= 0
    else { return nil }
    return RichOmittedPayload(bytes: bytes)
  }

  static func isSafe(_ block: RichContentBlock) -> Bool {
    guard case .image(let mime, let dataURL, _, _, _) = block,
      let classification = classify(dataURL)
    else { return false }
    return classification.mimeType.caseInsensitiveCompare(mime) == .orderedSame
  }

  private struct OptionalInteger { let value: Int64? }

  private static func optionalNonnegativeInteger(
    _ object: [String: RichJSON], _ key: String
  ) -> OptionalInteger? {
    guard let raw = object[key] else { return OptionalInteger(value: nil) }
    guard raw != .null, let value = raw.exactInt64Value, value >= 0 else { return nil }
    return OptionalInteger(value: value)
  }
}

enum RichAttachmentError: String, Sendable, Equatable {
  case empty = "empty_attachment"
  case tooLarge = "attachment_too_large"
  case invalid = "invalid_attachment"
}

struct RichAttachmentDecision: Sendable, Equatable {
  let queryValid: Bool
  let bodyWithinLimit: Bool
  let accepted: Bool
  let error: RichAttachmentError?
}

enum RichAttachmentPolicy {
  static let maximumBytes: Int64 = 20 * 1_024 * 1_024
  /// Matches JavaScript/Zod string length semantics used by the wire boundary.
  static let maximumNameUTF16Units = 255

  static func evaluate(name: String, byteCount: Int64) -> RichAttachmentDecision {
    let queryValid = !name.isEmpty && name.utf16.count <= maximumNameUTF16Units
    let bodyWithinLimit = (0...maximumBytes).contains(byteCount)
    let accepted = queryValid && bodyWithinLimit && byteCount > 0
    let error: RichAttachmentError?
    if accepted {
      error = nil
    } else if !queryValid || byteCount < 0 {
      error = .invalid
    } else if !bodyWithinLimit {
      error = .tooLarge
    } else {
      error = .empty
    }
    return RichAttachmentDecision(
      queryValid: queryValid,
      bodyWithinLimit: bodyWithinLimit,
      accepted: accepted,
      error: error
    )
  }
}
